[LV] Consider interleaving when -enable-wide-lane-mask=true

kmclaughlin-arm opened the pull request:
Currently the only way to enable the use of wide active lane masks is to pass
-enable-wide-lane-mask and force both interleaving & tail-folding with additional
flags. This patch changes selectInterleaveCount to consider interleaving if wide
lane masks were requested, although the feature remains off by default.

kmclaughlin-arm committed:
[LV] Consider interleaving & tail-folding when -enable-wide-lane-mask=true

Currently the only way to enable the use of wide active lane masks is to pass
-enable-wide-lane-mask and force both interleaving & tail-folding with additional
flags. This patch changes selectInterleaveCount & preferPredicateOverEpilogue to
consider both interleaving and tail-folding if wide lane masks were requested,
although the feature remains off by default.

Basic cost model changes are also included which reduce the cost of the
get.active.lane.mask intrinsic when the return type would require splitting,
but we know the whilelo (predicate pair) instruction can be used.

david-arm commented on llvm/test/Analysis/CostModel/AArch64/sve-intrinsics.ll:
```diff
@@ -1,6 +1,8 @@
 ; NOTE: Assertions have been autogenerated by utils/update_analyze_test_checks.py
 ; RUN: opt < %s -passes="print<cost-model>" -cost-kind=all 2>&1 -disable-output -S -mtriple=aarch64--linux-gnu -mattr=+sve | FileCheck %s --check-prefix=CHECK-VSCALE-1
-; RUN: opt < %s -passes="print<cost-model>" -cost-kind=all 2>&1 -mcpu=neoverse-v1 -disable-output -S -mtriple=aarch64--linux-gnu -mattr=+sve | FileCheck %s --check-prefix=CHECK-VSCALE-2
+; RUN: opt < %s -passes="print<cost-model>" -cost-kind=all 2>&1 -mcpu=neoverse-v1 -disable-output -S -mtriple=aarch64--linux-gnu -mattr=+sve | FileCheck %s --check-prefix=CHECK-VSCALE-2 --check-prefix=CHECK-SVE
```
I think instead of two separate `--check-prefix` arguments you can just write `--check-prefixes=CHECK-VSCALE-2,CHECK-SVE`

kmclaughlin-arm committed:
- Removed tail-folding changes
- Moved cost-model related changes to new PR
- Added optsize test to sve-wide-lane-mask.ll

kmclaughlin-arm committed:
- Rebase & run update_test_checks.py on sve-wide-lane-mask.ll

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -249,6 +249,10 @@ static cl::opt<TailFoldingStyle> ForceTailFoldingStyle(
 
+cl::opt<bool> llvm::EnableWideActiveLaneMask(
+    "enable-wide-lane-mask", cl::init(false), cl::Hidden,
+    cl::desc("Enable use of wide get active lane mask instructions"));
```
nit: Perhaps just say 'Enable use of wide lane masks in tail-folded loops'?

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -1346,6 +1355,15 @@ class LoopVectorizationCostModel {
     return getTailFoldingStyle() != TailFoldingStyle::None;
   }
 
+  bool useWideActiveLaneMask() const {
```
For both of the new functions you've added can you add some brief '///' comments?

david-arm reviewed (commented):
Thanks for the changes, the new version looks a lot simpler now!

kmclaughlin-arm committed:
- Document new functions and rename preferPredicatedEpilogue -> preferPredicatedLoop

david-arm reviewed (approved):
LGTM!

fhahn reviewed (approved):
New version LGTM, thanks

kmclaughlin-arm committed:
- Change comment in selectInterleaveCount
